### PR TITLE
[#1067] Basic implementation of edge and bane tracking

### LINKED ISSUE #1067: Implement Bane & Edge tracking
There are a number of general edge and bane effects that the system can track without worrying about odd conditionals

- Bane/Edge to hit a target (e.g. "ability rolls made against them gain an edge")
- Bane/Edge on certain abilities (e.g. "Gain an edge on strikes")
- Bane/Edge on tests with a characteristic (e.g. "Gain an edge on Might tests")
- Bane/Edge on tests with a skill (e.g. "Gain an edge on on tests that use the Handle Animals skill")

This does not cover GM-adjudication effects like "edge on tests to hide or sneak" or "edge against demons, devils, undead, and creatures not native to your current world".

These values do not need to be configurable as part of the Hero/NPC sheet and can only exist as AE paths.

### PR DESCRIPTION
**Pull request by metrzero**
Closes #1067 

I currently append " (+1 Edge)" to the end of skills. This is probably more appropriately done with a new format string to give translators options on where to present the additional modifier. I'll make that change if the rest of this is OK.

The UI currently misbehaves when there are more than 2 global edges or banes, as the dropdown cannot represent them, and so it reverts to 0. I think it will be necessary to move away from a dropdown to some other control, but was asked to hold off making any UI changes beyond the labeling.

### CODE REVIEW

File: src/module/applications/apps/power-roll-dialog.mjs [hunk 1]
```diff
@@ -132,11 +132,30 @@ export default class PowerRollDialog extends RollDialog {
...
-      const { label, group } = list[value];
-      accumulator.push({ label, group: groups[group].label, value });
+      const { group } = list[value];
+      let { label } = list[value];
+      if (value in skillModifiers) {
+        const modifiers = [];
+
+        const edges = skillModifiers[value].edges;
+        const edgeName = edges === 1 ? "Edge" : "Edges";
+        if (edges > 0) modifiers.push(game.i18n.format("DRAW_STEEL.ROLL.Power.Modifier.Label", { number: `+${edges}`, mod: edgeName }));
```
Comment (JPMeehan) [inline src/module/applications/apps/power-roll-dialog.mjs]: For this and the names, I'm pretty sure there are more formal ways to leverage i18n pluralization rules

File: src/module/data/actor/base.mjs [hunk 2]
```diff
@@ -396,6 +396,38 @@ export default class BaseActorModel extends DrawSteelSystemModel {
    * @returns {Promise<DrawSteelChatMessage | null>}
    */
   async rollCharacteristic(characteristic, options = {}) {
+    const skillModifiers = {};
+
+    // Extract applicable modifier bonuses
+    for (const bonus of this._abilityBonuses) {
```
Comment (JPMeehan) [inline src/module/data/actor/base.mjs]: I don't think I'd be expecting these to be Ability bonuses, just normal AEs modifying actor data

File: src/module/applications/apps/power-roll-dialog.mjs [hunk 3]
```diff
@@ -132,11 +132,34 @@ export default class PowerRollDialog extends RollDialog {
    */
   _prepareSkillOptions(context) {
     const { list, groups } = ds.CONFIG.skills;
+    const skillModifiers = context.skillModifiers;
+
+    const pr = new Intl.PluralRules(game.i18n.lang, { type: "cardinal" });
```
Comment (JPMeehan) [inline src/module/applications/apps/power-roll-dialog.mjs]: I wonder if we can cache this, e.g. make it a static property

File: src/module/data/item/ability.mjs [hunk 4]
```diff
@@ -524,6 +524,23 @@ export default class AbilityModel extends BaseItemModel {
 
     // Restrained conditions check
     if (this.actor.statuses.has("restrained")) options.modifiers.banes += 1;
+
+    // Apply active effect modifiers from abilityModifier effects
+    for (const bonus of (this.actor.system._abilityBonuses ?? [])) {
+      if (!bonus.key.startsWith("power.")) continue;
+      if (!bonus.filters.keywords.isSubsetOf(this.keywords)) continue;
```
Comment (JPMeehan) [inline src/module/data/item/ability.mjs]: Shouldn't the looping be handled in _applyAbilityBonuses where we're already looping and checking filters? Then on line 428 you can initialize banes & edges to `this.power.roll.banes` etc.

File: src/module/applications/apps/power-roll-dialog.mjs [hunk 5]
```diff
@@ -132,11 +141,37 @@ export default class PowerRollDialog extends RollDialog {
    */
   _prepareSkillOptions(context) {
     const { list, groups } = ds.CONFIG.skills;
+    const skillModifiers = context.skillModifiers;
+
+    if (!this.constructor.EdgePluralFormatter) {
+      this.constructor.EdgePluralFormatter = new Intl.PluralRules(game.i18n.lang, { type: "cardinal" });
+    }
+    const pr = this.constructor.EdgePluralFormatter;
```
Comment (JPMeehan) [inline src/module/applications/apps/power-roll-dialog.mjs]: Null coalescing here should be sufficient?
```suggestion
    const pr = this.constructor.EdgePluralFormatter ??= new Intl.PluralRules(game.i18n.lang, { type: "cardinal" });
```

File: src/module/applications/apps/power-roll-dialog.mjs [hunk 6]
```diff
@@ -31,6 +31,15 @@ export default class PowerRollDialog extends RollDialog {
 
   /* -------------------------------------------------- */
 
+  /**
+   * Formatter to select the correct plural form for "Edge".
+   * Must be initialized before first use to ensure language is available.
+   * @type {Intl.PluralRules | null}
+   */
+  static EdgePluralFormatter = null;
```
Comment (krbz999) [inline src/module/applications/apps/power-roll-dialog.mjs]: I think we can make this a getter. Or perhaps we want to extend `game.i18n`? Perhaps that's insane.

Comment (JPMeehan) [reply src/module/applications/apps/power-roll-dialog.mjs]: v14 looks like they're just making local cached versions of formatters (e.g. Active Effect duration nonsense) which is why that was my suggestion.

File: src/module/data/actor/base.mjs [hunk 7]
```diff
@@ -140,6 +140,13 @@ export default class BaseActorModel extends DrawSteelSystemModel {
       // Can consider removing in v14 after phases are introduced
       multiplier: 1,
     });
+
+    Object.values(this.characteristics).forEach((chr) => {
```
Comment (JPMeehan) [inline src/module/data/actor/base.mjs]: Stylistically prefer `for of` loops
```suggestion
    for (const chr of Object.values(this.characteristics)) {
```

File: src/module/data/item/ability.mjs [hunk 8]
```diff
@@ -201,6 +201,19 @@ export default class AbilityModel extends BaseItemModel {
...
+      if (bonus.key.startsWith("power.")) {
+        switch (bonus.key) {
+          case "power.roll.banes":
+            this.power.roll.banes = this.power.roll.banes ?? 0 + (Number(bonus.value) || 0);
+            break;
+          case "power.roll.edges":
+            this.power.roll.edges = this.power.roll.edges ?? 0 + (Number(bonus.value) || 0);
+            break;
+          default:
+            break;
```
Comment (JPMeehan) [inline src/module/data/item/ability.mjs]: Can just leave a blank default case off
```suggestion
```

**Review by JPMeehan — COMMENTED** [review]
Ultimately a bunch of stylistic notes & nitpicks, I think this PR is close to done.

**Comment by metrzero** [comment]
I'm out of town for a few more days, but should be able to get to it this coming weekend.

**Comment by metrzero** [comment]
I should be able to get to it today/tomorrow. Thanks for the reminder.

**Review by JPMeehan — APPROVED** [review]
Great, thanks for tackling this, it was a big project
